### PR TITLE
Improve login page

### PR DESCRIPTION
**Pull request by KTibow**
## Proposed change
The current login page uses a lot of external components which makes it have a large bundle size. This PR rewrites the login page, which brings all of these benefits:
- Feels snappier
  - tinier bundle sizes: 278.7kb -> 103.1kb
  - I try to update the UI as soon as possible instead of waiting for the next thing to load
- UI that looks better and is easier to understand (removed redundant descriptions, clarified what start over does, etc)
- Fixes a number of other random things
  - marking required fields as required
  - using number fields for one time auth codes
  - bigger show/hide password touch target
  - easier to understand code

While I understand it may take a bit to review all the code, I'll try to be as receptive as possible to feedback, and hope that this can get merged soon.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
```yaml
# If you want a whole bunch of auth providers to test with
homeassistant:
  auth_mfa_modules:
    - type: totp
    - type: notify
      include:
        - mobile_app_gm1917
  auth_providers:
    - type: homeassistant
    - type: trusted_networks
      trusted_networks:
        - 127.0.0.1
        - ::1
        - 192.168.1.0/24
    - type: legacy_api_password
      api_password: "hi"
```

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

Relevant issues: #10274

### CODE REVIEW

**Comment by steverep** [comment]
> no issues, discussions, or documentation are relevant here

#10274 and its predecessors are highly relevant here.  I would encourage you to read through that for a lot of historical context if you haven't already.

> The current login page uses Polymer components, which break password managers,

They use lit, not polymer, which is much smaller and faster.  The password manager breakage is self-inflicted as many did not support shadow roots until recently - see https://github.com/home-assistant/frontend/issues/10274#issuecomment-1670141427 for recent statuses.  For those that don't, we have the polyfill, which just got an upgrade to improve support in #17830.

> have weird focus states, and are slow to load. This rewrites the login page using Vanilla JS™️, with all these benefits:

This reasoning is misplaced IMO, because the same comment should apply to the entire UI since they use the exact same components.  If there are ways to improve speed and focus appearance, I'm all for it, but it should be applied universally.  FYI, Material 3 components have much better focus appearance, which HA will switch to when they are fully released I'm sure.

> Fixes a number of other random things (eg marking required fields as required, using number fields for one time auth codes)

They are indeed already marked as required so I'm confused by that point.  Also, although many implementations use only numbers, two-factor codes do not have to be numeric.

**Comment by KTibow** [comment]
> I would encourage you to read through that for a lot of historical context if you haven't already.

> They use lit, not polymer, which is much smaller and faster.

Yes - still, they're components. Eg `custom-style` is 47.6kb of weight, and the rest of components (eg `mwc-button`) are decently heavy. Just using vanilla HTML/JS prevents all these problems in the first place.

> This reasoning is misplaced IMO, because the same comment should apply to the entire UI since they use the exact same components. If there are ways to improve speed and focus appearance, I'm all for it, but it should be applied universally. FYI, Material 3 components have much better focus appearance, which HA will switch to when they are fully released I'm sure.

I'm working on other ways of decreasing bundle size - as for focus, yeah I'll leave that up to Material 3. I see the login page as a special case, as it's inconvenient to wait, login, wait, and view HA, and it has such simple inputs/flow that it's possible to not use components.

> They are indeed already marked as required so I'm confused by that point.

Don't think so :wink:
[Screencast from 2023-09-24 13-54-08.webm](https://github.com/home-assistant/frontend/assets/10727862/dc704e8b-ff76-4187-9342-44eaf23cbe6f)

> Also, although many implementations use only numbers, two-factor codes do not have to be numeric.

Oh. Can you send some `home-assistant/core` links to other 2FA implementations that aren't numeric?

**Comment by steverep** [comment]
> I'm working on other ways of decreasing bundle size - as for focus, yeah I'll leave that up to Material 3. I see the login page as a special case, as it's inconvenient to wait, login, wait, _then_ view HA, and it has such simple inputs/flow that it's possible to not use components.

But you're just exchanging a little less time in the first wait for more time in the second since a few less assets are loaded after login.

> Don't think so 😉 

I'm blind so I do not see what you are trying to point out, but the `input` elements clearly have the `required` attribute.  If that's not clear visually, that should be raised as a separate issue.  And again, the fix should be applied universally so all required fields get the benefit.

> Oh. Can you send some `home-assistant/core` links to other 2FA implementations that aren't numeric?

That wasn't my point - just that from a general validation perspective it's not a hard requirement.  HA could certainly tie the input mode to the 2FA scheme being used, but again that's a separate issue I think.  It would also need to be applied during 2FA setup.

**Comment by KTibow** [comment]
> But you're just exchanging a little less time in the first wait for more time in the second since a few less assets are loaded after login.

Not entirely sure what you mean. JS isn't reused across login and the app as it is right now.

> I'm blind so I do not see what you are trying to point out, but the input elements clearly have the required attribute.

If you try to submit a input field that's required but empty, it should prevent you from submitting it. It doesn't do that as is and does do that in my version.

**Comment by piitaya** [comment]
Hello 👋
Changing the UX/UI and fixing the password manager support should be done in separate PRs.

For UX/UI, the changes must be discussed with the design team (@matthiasdebaat and @Madelena). 

Some small changes has already been done with https://github.com/home-assistant/frontend/pull/17459 to simplify the UI but we have no short plan to migrate to Material 3 because there is still a lot of breaking change between each release and no stable release is planned on the Material Web library (https://github.com/material-components/material-web). We will consider the migration once it's more stable.
Also, removing the logo should be discussed. I don't really understand why it's replaced by a lock icon. It's a very common pattern to display the logo on the login page.

For the technical perspective, we must not go out of `lit`. We don't want to go backwards with manual templating, manual dom manipulation or functions in the window object. It looks very hacky. Maintenance of this page will be hard with these changes.
We can use lit to render html outside of shadow dom using `createRenderRoot`. This can also be a solution to password manager issue. But we would to rewrite part of `ha-textfield`, `ha-form` and `ha-auth-flow`.

Lot of work has been done by @steverep to improve password manager support (last PR : https://github.com/home-assistant/frontend/pull/17830). The only issue remaining (correct me if I'm wrong Steve) is the icon not showing for autocomplete.

For big changes like this in the future (this change included), I advice you to open a front-end or architecture discussion because there are many implications (e.g. maintenance, legacy support, historical discussions).

**Comment by KTibow** [comment]
> Changing the UX/UI and fixing the password manager support should be done in separate PRs.

> Lot of work has been done by @steverep to improve password manager support (last PR : https://github.com/home-assistant/frontend/pull/17830). The only issue remaining (correct me if I'm wrong Steve) is the icon not showing for autocomplete.

This doesn't really make since when the goal is to use pure HTML to fix password manager support, and I would rather make a new UI than have to make one PR that clones the old UI and another PR that adds the new UI

> Also, removing the logo should be discussed. I don't really understand why it's replaced by a lock icon. It's a very common pattern to display the logo on the login page.

Makes it look better (IMO)

> For the technical perspective, we must not go out of lit. We don't want to go backwards with manual templating, manual dom manipulation or functions in the window object. It looks very hacky. Maintenance of this page will be hard with these changes.
We can use lit to render html outside of shadow dom using createRenderRoot. This can also be a solution to password manager issue. But we would to rewrite part of ha-textfield, ha-form and ha-auth-flow.

Fair enough. I can rewrite this in Lit. Only catch is that as soon as I add anything from `@material`, it's a chonky web component and uses a shadow root.

> We recently updated the onboarding. For this design change, we need to follow that in terms of layout, logo, positions, buttons and animated background

I don't get this stance - the onboarding should be a different experience than a login page.
> For big changes like this in the future (this change included), I advice you to open a front-end or architecture discussion because there are many implications (e.g. maintenance, legacy support, historical discussions).

Discussions (and `#devs-frontend`) always seem like I'm screaming into a void

To summarize:
- I want pure HTML, but I can do pure, shadowroot-less Lit. Is that okay?
- I would rather not split up the rewrite of UI and rewrite of dom manipulation. Is that okay?

**Comment by bramkragten** [comment]
Check this POC from @piitaya : <https://github.com/home-assistant/frontend/pull/18015>

**Comment by KTibow** [comment]
Cool - however that still doesn't fix most of the problems this one addresses. Hopefully I can push a Lit version today.

**Comment by bramkragten** [comment]
You mean the bundle size, etc? I think the better way is to merge the PR of @piitaya, and then you can create a new PRs with improvements. But I think it is wise to discuss this before you make them.

**Comment by KTibow** [comment]
I converted this PR to use Lit, and I'd say it's easier to read than the previous code (at least to me). I'll update the above PR description with more info once I finish school today.

**Comment by frenck** [comment]
> (Also I kinda did [discuss this](https://community.home-assistant.io/t/feedback-needed-redesigned-the-login-page-what-do-you-think/617411?u=ktibow) already)

That is the community forum, which is a general community support place. If you want to discuss architecture, please use our architectural repository instead.

**Comment by KTibow** [comment]
I think I messed up my rebase

**Comment by KTibow** [comment]
> I think the better way is to merge the PR of @piitaya, and then you can create new PRs with improvements.

piitaya's PR got merged. After some rebasing, this PR is now the "new PR with improvements", and I'm eager to respond to some new reviews (or any old reviews that I didn't address).

File: src/html/authorize.html.template [hunk 1]
```diff
@@ -1,69 +1,96 @@
...
     <%= renderTemplate("_preload_roboto.html.template") %>
     <script crossorigin="use-credentials">
       // Safari 12 and below does not have a compliant ES2015 implementation of template literals, so we ship ES5
-      if (!isS11_12) {
+      window.latestJS = window.latestJS || !isS11_12;
+      if (window.latestJS) {
         <% for (const entry of latestEntryJS) { %>
           import("<%= entry %>");
         <% } %>
-        window.latestJS = true;
```
Comment (bramkragten) [inline src/html/authorize.html.template:66]: We are testing here if `import()` will work, if it throws it doesnt get here so `latestJS` is not set to true.

With your code `latestJS` is always set to `true` except when the browser is Safari 11/12?

File: src/entrypoints/authorize.ts [hunk 2]
```diff
@@ -1,10 +1,2 @@
-// Compat needs to be first import
-import "../resources/compatibility";
```
Comment (bramkragten) [inline src/entrypoints/authorize.ts:2]: Why would this not be needed anymore? This is needed for the ES5 build.

File: src/auth/ha-authorize.ts [hunk 3]
```diff
@@ -1,251 +1,516 @@
...
+    for (const provider of this._authProviders) {
+      if (provider === this._authProvider) continue;
+      // eslint-disable-next-line @typescript-eslint/no-loop-func
+      const useOtherMethod = async () => {
+        const result = await loadStep(
+          this._step,
+          clientId,
+          provider,
+          redirectUri
+        );
```
Comment (bramkragten) [inline src/auth/ha-authorize.ts:192]: We don't use inline event handlers, this should be a class function

File: src/auth/ha-authorize.ts [hunk 4]
```diff
@@ -1,251 +1,516 @@
...
-      // Forward to main screen which will redirect to right onboarding page.
-      if (
-        response.status === 400 &&
-        authProviders.code === "onboarding_required"
-      ) {
-        location.href = `/onboarding.html${location.search}`;
-        return;
-      }
+const styleTag = document.createElement("style");
+styleTag.innerHTML = `
```
Comment (bramkragten) [inline src/auth/ha-authorize.ts:342]: Don't use `innerHTML`, use `innerText`

File: src/auth/ha-authorize.ts [hunk 5]
```diff
@@ -1,251 +1,516 @@
...
+input[type=checkbox] {
+  width: 18px;
+  height: 18px;
+  margin: 0 16px 0 0;
+  cursor: pointer;
+}
 
-  private async _handleAuthProviderPick(ev) {
-    this._authProvider = ev.detail;
+ha-authorize-field + .buttons {
```
Comment (bramkragten) [inline src/auth/ha-authorize.ts:487]: Keep the styling of elements with those elements instead of here and the template

**Comment by bramkragten** [comment]
In general apart from the design itself, which @matthiasdebaat probably has some ideas for, I don't think we want to go in a direction that we create our own UI elements, yes these elements are smaller, but they also lack a lot of the functionality of the elements we currently use, like hover/active states, animations, labels, etc. The reason we use material elements is so we don't have to worry about all of that, and for example accessibility, and don't have to maintain it.

File: src/resources/roboto.js
Comment (steverep) [inline src/resources/roboto.js:1]: Why just remove comments?

File: src/auth/ha-authorize.ts (hunk 3 above)
Comment (KTibow) [reply src/auth/ha-authorize.ts:192]: It's not that simple. Each button performs a different function. While I could do something like moving the core to the class and making a wrapper function, binding the function, or identifying the auth provider from the element's text, I think it's the most reasonable to do it this way.

File: src/resources/roboto.js
Comment (KTibow) [reply src/resources/roboto.js:1]: Because I was going through the code when I was first trying to understand the complicated `ha-authorize` component [and fixed that](https://github.com/home-assistant/frontend/pull/18005/commits/f81306e669ad8cea4abafc7bc7bab6577f2c0233)

File: src/auth/ha-authorize.ts (hunk 5 above)
Comment (KTibow) [reply src/auth/ha-authorize.ts:487]: > with those elements

What would this look like? Something like a const at the top level and a `<style>${styles}</style>` at each render return, or did you mean something else? (Also you could argue having all the styles in one place makes more sense when they aren't scoped)

> and the template

Most of the template styling is there to prevent a layout shift, so everything except the `ha-authorize-field` styling isn't going anywhere.

**Comment by KTibow** [comment]
> The reason we use material elements is so we don't have to worry about all of that, and for example accessibility, and don't have to maintain it.

Hmm. Here are some reasons to still not use material elements:
1. There are some problems with the existing implementation, like allowing to submit an empty required field or having a tiny touch target for the show/hide password button
1. The way we currently use styling for the input fields is weird
1. AFAIK the inputs do everything they needs to do already
1. I can maintain it myself

For now I'm just going to keep addressing as many reviews as I can -- either this PR gets merged (which doesn't feel likely) or it goes stale/is denied, in which case I'll <details><summary>split this PR up</summary>
1. Restructure data flow
2. Adjust the buttons/choosers/wording to be easier to navigate
3. Improve the design
4. Use custom buttons
5. Use custom inputs</details> so the less objectionable parts can be merged.

**Comment by KTibow** [comment]
Closing so I can split this PR up
